tests/functional/object/mcg/test_bucket_replication_with_versioning.py

```diff
@@ -46,103 +50,312 @@ def reduce_replication_delay_setup(self, add_env_vars_to_noobaa_core_class):
-                    )
-                    break
-                logger.warning(
-                    f"Source and target etags do not match: {source_etags} != {target_etags}"
+        assert (
+            len(source_versions) == 1
+        ), f"Expected a single version on a suspended bucket, got: {source_versions}"
+        wait_for_object_versions_match(
+            mcg_obj,
+            awscli_pod,
+            source_bucket.name,
+            target_bucket.name,
+            obj_key=obj_key,
+        )
+
+        # 6. Enable versioning on the source bucket and suspend the versioning on the target bucket
+        put_bucket_versioning_via_awscli(
+            mcg_obj, awscli_pod, source_bucket.name, status="Enabled"
+        )
+        put_bucket_versioning_via_awscli(
+            mcg_obj, awscli_pod, target_bucket.name, status="Suspended"
+        )
+
+        # 7. Write versions to the source bucket under a different object key
+        obj_key = "test_obj_" + str(uuid4())[:4]
+        upload_obj_versions(
+            mcg_obj,
+            awscli_pod,
+            source_bucket.name,
+            obj_key=obj_key,
+            amount=versions_amount,
+        )
+
+        # 8. Wait to verify that only the latest version was uploaded to the target bucket
+        source_versions = get_obj_versions(
+            mcg_obj, awscli_pod, source_bucket.name, obj_key
+        )
+        latest_version_etag = source_versions[0]["ETag"]
+
+        for target_versions in TimeoutSampler(
+            timeout=300,
+            sleep=30,
+            func=get_obj_versions,
+            mcg_obj=mcg_obj,
+            awscli_pod=awscli_pod,
+            bucket_name=target_bucket.name,
+            obj_key=obj_key,
+        ):
+            if len(target_versions) > 1:
+                raise UnexpectedBehaviour(
+                    f"Expected a single version, got: {target_versions}"
                 )
-                last_target_etags = target_etags
-        except TimeoutExpiredError as e:
-            err_msg = (
-                f"Source and target etags do not match after {TIMEOUT} seconds:\n"
-                f"Source etags:\n{json.dumps(source_etags, indent=2)}\n"
-                f"Target etags:\n{json.dumps(last_target_etags, indent=2)}"
+            target_version_etag = (
+                target_versions[0].get("ETag")
+                if target_versions and isinstance(target_versions[0], dict)
+                else None
             )
-            logger.error(err_msg)
-            raise TimeoutExpiredError(f"{str(e)}\n\n{err_msg}")
+            if target_version_etag == latest_version_etag:
+                logger.info("Only the later version was replicated as expected")
+                break
+            else:
+                logger.warning(
+                    f"Expected the latest version {latest_version_etag}, got: {target_version_etag}"
+                )
```

Comment:
We should add assert here right @sagihirshfeld , Please correct me if I am wrong

Reply:
We do not need to assert here - if we won't reach the desired state within the timeout, then TimeoutSampler will fail the test with a TimeoutExpiredError exception.

I could wrap this entire section in a try-except block in order to catch that exception and then fail the test with a more meaningful message, but I think this part is too messy as it is already.

Reply:
Thanks for the clarification